### PR TITLE
OpenGL support for the GBM windowing system

### PR DESCRIPTION
**Pull request by a1rwulf**
## Description
This adds support for OpenGL rendering to the GBM platform.
I have added `-DGBM_RENDER_SYSTEM` which can be either `gl` or `gles` like
we already have for wayland.
Sample command to compile:
`cmake -DENABLE_OPENGLES=OFF -DENABLE_OPENGL=ON -DGBM_RENDER_SYSTEM=gl -DCORE_PLATFORM_NAME=gbm`

## Motivation and Context
I did this mainly for my usage of the Intel NUC which is capable of OpenGL.
Furthermore I think it helps us not degrading our architecture by binding other code segments
directly to OpenGL ES, while they should only be bound to EGL/GBM.

## How Has This Been Tested?
I did runtime tests (gui,video) on Ubuntu 18.04 Intel NUC.
@garbear I did not (yet) run games as my setup is incomplete and I have no games stuff compiled.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Cosmetic change (non-breaking change that doesn't touch code)
- [ ] None of the above (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

@Kwiboo If you have some time, please check if your stuff still works.
@lrusak @garbear Maybe one of you can run a retroplayer test?

Todo:
- [x] Add new cmake flag to our docs

### CODE REVIEW

File: xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGL.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
+ /*
```
Comment (hudokkow) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGL.cpp]: Stray space before `/*`

File: xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGLES.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
+ /*
```
Comment (hudokkow) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGLES.cpp]: Stray space before `/*`

File: xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGLES.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
+ /*
```
Comment (hudokkow) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGLES.h]: Stray space before `/*`

File: xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (hudokkow) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp]: Missing opening `/*`

File: xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
+/*
```
Comment (hudokkow) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.h]: Missing opening `/*`

File: xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGLES.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
+/*
```
Comment (hudokkow) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGLES.cpp]: Missing opening `/*`

File: xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp (hunk 4 above)
Comment (hudokkow) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp]: Actually, only `/` is missing

**Comment by hudokkow** [comment]
Thanks for using SPDX License Identifier!

Please adapt build instructions under https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md#41-configure-build (see wayland example)

**Comment by wsnipex** [comment]
works great for me on my ivy bridge i3 :+1:

File: xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    CreateTexture();
+
+  glBindTexture(m_textureTarget, m_textureId);
+
+  const int stride = GetFrameSize() / m_height;
+
+  glPixelStorei(GL_UNPACK_ALIGNMENT, m_bpp);
+
+  glPixelStorei(GL_UNPACK_ROW_LENGTH, stride / m_bpp);
+  glTexSubImage2D(m_textureTarget, 0, 0, 0, m_width, m_height, m_pixelformat, m_pixeltype, m_data.data());
```
Comment (FernetMenta) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp]: you should really consider using PBOs here. OpenGL rendering can't be done worse than with this code, by using glTexSubImage2D

Comment (a1rwulf) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp]: @FernetMenta I agree.
But would you mind me getting it in as is and improve the renderer in a second iteration?

Comment (FernetMenta) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp]: Sure. I would place a comment in. This is a big change and history is hard to follow.

Comment (a1rwulf) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp]: Yep, will do that.

File: xbmc/windowing/gbm/WinSystemGbmGLContext.cpp [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  m_eglContext.SwapBuffers();
+
+  CWinSystemGbm::SetFullScreen(fullScreen, res, blankOtherDisplays);
+  CRenderSystemGL::ResetRenderSystem(res.iWidth, res.iHeight);
+
+  if (!m_delayDispReset)
+  {
+    CSingleLock lock(m_resourceSection);
+
+    for (std::vector<IDispResource *>::iterator i = m_resources.begin(); i != m_resources.end(); ++i)
```
Comment (Rechi) [inline xbmc/windowing/gbm/WinSystemGbmGLContext.cpp]: range based for loop

File: xbmc/windowing/gbm/WinSystemGbmGLContext.cpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+  {
+    CWinSystemGbm::WaitVBlank();
+  }
+
+  if (m_delayDispReset && m_dispResetTimer.IsTimePast())
+  {
+    m_delayDispReset = false;
+    CSingleLock lock(m_resourceSection);
+
+    for (std::vector<IDispResource *>::iterator i = m_resources.begin(); i != m_resources.end(); ++i)
```
Comment (Rechi) [inline xbmc/windowing/gbm/WinSystemGbmGLContext.cpp]: range based for loop

File: xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.h [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "video/RenderBufferSysMem.h"
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.h]: Can you use paths relative to the root xbmc/ folder for non-same-directory files?

**Review by garbear — APPROVED** [review]
RP changes look good

File: xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGLES.cpp [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    m_pixelformat = GL_RGB;
+    m_bpp = sizeof(uint16_t);
+    return true;
+  }
+  default:
+    break;
+  }
+
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (a1rwulf) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferPoolOpenGLES.cpp]: Should be addressed as well, but it's hard to see in gh.

**Comment by lrusak** [comment]
Why do you not want to use the gbm renderer for retroplayer?

**Comment by a1rwulf** [comment]
> Why do you not want to use the gbm renderer for retroplayer?

Because it uses a TextureTarget that is not available for OpenGL:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/RetroPlayer/process/gbm/RenderBufferGBM.h#L51

**Comment by FernetMenta** [comment]
OpenGL can use PBOs which is simpler and more robust

**Comment by Rechi** [comment]
The toolchain change either has to be squashed with the commit that adds the requirement for setting `GBM_RENDER_SYSTEM` or at least committed before it.

**Comment by a1rwulf** [comment]
OK got rid of the introduced warnings and squashed toolchain commit into 88a6381.
Let's see if jenkins is happy now.

**Comment by a1rwulf** [comment]
@Rechi I've addressed the warnings and your request.
@lrusak if you're also good with the gbm part I'd like to merge soon.

**Comment by garbear** [comment]
Can you make GL and GLES derive from a GLBase class? totally cool if it's in a follow up PR.

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/CMakeLists.txt [hunk 12]
```diff
@@ -49,7 +49,7 @@ if(CORE_SYSTEM_NAME STREQUAL android)
                       RendererMediaCodecSurface.h)
 endif()
 
-if(CORE_PLATFORM_NAME_LC STREQUAL gbm)
+if(CORE_PLATFORM_NAME_LC STREQUAL gbm AND OPENGLES_FOUND)
```
Comment (Kwiboo) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/CMakeLists.txt]: `RendererDRMPRIME.cpp` should work with OpenGL not just OpenGLES but guess there is only GLES device/platforms that is using the direct-to-plane renderer at the moment

Comment (a1rwulf) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/CMakeLists.txt]: If you don't mind I'd rather keep it as is until we really need it for OpenGL.

**Comment by a1rwulf** [comment]
> Can you make GL and GLES derive from a GLBase class in a follow up PR? No hurry.

I'm going to do more renderer improvements once this is in and I'm happy to have a look at it as discussed.
